File: python/src/wgrok/webex.py
"""Thin Webex REST client for sending messages, cards, and fetching resources."""

from __future__ import annotations

import asyncio

import aiohttp

WEBEX_API_BASE = "https://webexapis.com/v1"
MAX_RETRIES = 3
WEBEX_MESSAGES_URL = f"{WEBEX_API_BASE}/messages"
WEBEX_ATTACHMENT_ACTIONS_URL = f"{WEBEX_API_BASE}/attachment/actions"

ADAPTIVE_CARD_CONTENT_TYPE = "application/vnd.microsoft.card.adaptive"


def _headers(token: str) -> dict:
    return {"Authorization": f"Bearer {token}", "Content-Type": "application/json"}


async def _manage_session(session, func):
    """Run func with session, closing it afterward if we created it."""
    owns_session = session is None
    if owns_session:
        session = aiohttp.ClientSession()
    try:
        return await func(session)
    finally:
        if owns_session:
            await session.close()


async def _request_with_retry(session, method, url, headers, **kwargs):
    """Execute an HTTP request with Retry-After handling for 429 responses."""
    for attempt in range(MAX_RETRIES + 1):
        async with session.request(method, url, headers=headers, **kwargs) as resp:
            if resp.status == 429:
                if attempt >= MAX_RETRIES:
                    resp.raise_for_status()
                try:
                    retry_after = min(int(resp.headers.get("Retry-After", "1")), 300)
                except (ValueError, TypeError):
                    retry_after = 1
                await asyncio.sleep(retry_after)
                continue
            resp.raise_for_status()
            return await resp.json()
    # Should not reach here, but just in case
    msg = f"HTTP request to {url} failed after {MAX_RETRIES + 1} attempts"
    raise aiohttp.ClientResponseError(None, None, message=msg, status=429)


async def send_message(
    token: str,
    to_email: str,
    text: str,
    session: aiohttp.ClientSession | None = None,
    room_id: str = "",
) -> dict:
    """Send a text-only Webex message to a person by email or room by ID."""
    payload = {"roomId": room_id, "text": text} if room_id else {"toPersonEmail": to_email, "text": text}

    async def _do(s):
        return await _request_with_retry(s, "POST", WEBEX_MESSAGES_URL, _headers(token), json=payload)

    return await _manage_session(session, _do)


async def send_card(
    token: str,
    to_email: str,
    text: str,
    card: dict,
    session: aiohttp.ClientSession | None = None,
    room_id: str = "",
) -> dict:
    """Send a Webex message with an adaptive card attachment.

    Args:
        text: Fallback text for clients that can't render cards.
        card: Adaptive Card JSON body (the content inside the attachment).
        room_id: Optional room ID to send to room instead of person email.
    """
    if room_id:
        payload = {
            "roomId": room_id,
            "text": text,
            "attachments": [{"contentType": ADAPTIVE_CARD_CONTENT_TYPE, "content": card}],
        }
    else:
        payload = {
            "toPersonEmail": to_email,
            "text": text,
            "attachments": [{"contentType": ADAPTIVE_CARD_CONTENT_TYPE, "content": card}],
        }

    async def _do(s):
        return await _request_with_retry(s, "POST", WEBEX_MESSAGES_URL, _headers(token), json=payload)

    return await _manage_session(session, _do)


async def get_message(
    token: str,
    message_id: str,
    session: aiohttp.ClientSession | None = None,
) -> dict:
    """Fetch full message details by ID (includes attachments)."""
    url = f"{WEBEX_MESSAGES_URL}/{message_id}"

    async def _do(s):
        return await _request_with_retry(s, "GET", url, _headers(token))

    return await _manage_session(session, _do)


async def get_attachment_action(
    token: str,
    action_id: str,
    session: aiohttp.ClientSession | None = None,
) -> dict:
    """Fetch an attachment action (card submission) by ID."""
    url = f"{WEBEX_ATTACHMENT_ACTIONS_URL}/{action_id}"

    async def _do(s):
        return await _request_with_retry(s, "GET", url, _headers(token))

    return await _manage_session(session, _do)


def extract_cards(message: dict) -> list[dict]:
    """Extract adaptive card content dicts from a message's attachments."""
    attachments = message.get("attachments") or []
    return [
        att["content"]
        for att in attachments
        if att.get("contentType") == ADAPTIVE_CARD_CONTENT_TYPE and "content" in att
    ]

File: go/webex.go
package wgrok

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"time"
)

const (
	WebexAPIBase            = "https://webexapis.com/v1"
	AdaptiveCardContentType = "application/vnd.microsoft.card.adaptive"
	MaxRetries              = 3
)

// Package-level URLs, overridable in tests.
var (
	WebexMessagesURL          = WebexAPIBase + "/messages"
	WebexAttachmentActionsURL = WebexAPIBase + "/attachment/actions"
)

// CardAttachment represents a Webex message card attachment.
type CardAttachment struct {
	ContentType string      `json:"contentType"`
	Content     interface{} `json:"content"`
}

type sendMessagePayload struct {
	ToPersonEmail string           `json:"toPersonEmail"`
	Text          string           `json:"text"`
	Attachments   []CardAttachment `json:"attachments,omitempty"`
}

// SendMessage sends a text-only Webex message to a person by email.
func SendMessage(token, toEmail, text string, client *http.Client) (map[string]interface{}, error) {
	payload := sendMessagePayload{ToPersonEmail: toEmail, Text: text}
	return postMessage(token, payload, client)
}

// SendCard sends a Webex message with an adaptive card attachment.
func SendCard(token, toEmail, text string, card interface{}, client *http.Client) (map[string]interface{}, error) {
	payload := sendMessagePayload{
		ToPersonEmail: toEmail,
		Text:          text,
		Attachments: []CardAttachment{
			{ContentType: AdaptiveCardContentType, Content: card},
		},
	}
	return postMessage(token, payload, client)
}

func postMessage(token string, payload sendMessagePayload, client *http.Client) (map[string]interface{}, error) {
	if client == nil {
		client = http.DefaultClient
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("marshal message payload: %w", err)
	}

	for attempt := 0; attempt <= MaxRetries; attempt++ {
		req, err := http.NewRequest("POST", WebexMessagesURL, bytes.NewReader(body))
		if err != nil {
			return nil, err
		}
		req.Header.Set("Authorization", "Bearer "+token)
		req.Header.Set("Content-Type", "application/json")

		resp, err := client.Do(req)
		if err != nil {
			return nil, fmt.Errorf("send message: %w", err)
		}
		defer resp.Body.Close()

		// Check for 429 Too Many Requests
		if resp.StatusCode == http.StatusTooManyRequests {
			if attempt < MaxRetries {
				// Read Retry-After header, default to 1 second
				retryAfter := 1
				if retryAfterStr := resp.Header.Get("Retry-After"); retryAfterStr != "" {
					if parsed, err := strconv.Atoi(retryAfterStr); err == nil {
						retryAfter = parsed
					}
				}
				time.Sleep(time.Duration(retryAfter) * time.Second)
				continue
			}
			// Max retries exhausted, fall through to readJSONResponse which will handle the error
		}

		return readJSONResponse(resp)
	}
	return nil, fmt.Errorf("unreachable")
}

// GetMessage fetches full message details by ID (includes attachments).
func GetMessage(token, messageID string, client *http.Client) (map[string]interface{}, error) {
	url := WebexMessagesURL + "/" + messageID
	return getJSON(token, url, client)
}

// GetAttachmentAction fetches an attachment action (card submission) by ID.
func GetAttachmentAction(token, actionID string, client *http.Client) (map[string]interface{}, error) {
	url := WebexAttachmentActionsURL + "/" + actionID
	return getJSON(token, url, client)
}

func getJSON(token, url string, client *http.Client) (map[string]interface{}, error) {
	if client == nil {
		client = http.DefaultClient
	}

	for attempt := 0; attempt <= MaxRetries; attempt++ {
		req, err := http.NewRequest("GET", url, nil)
		if err != nil {
			return nil, err
		}
		req.Header.Set("Authorization", "Bearer "+token)
		req.Header.Set("Content-Type", "application/json")

		resp, err := client.Do(req)
		if err != nil {
			return nil, fmt.Errorf("GET %s: %w", url, err)
		}
		defer resp.Body.Close()

		// Check for 429 Too Many Requests
		if resp.StatusCode == http.StatusTooManyRequests {
			if attempt < MaxRetries {
				// Read Retry-After header, default to 1 second
				retryAfter := 1
				if retryAfterStr := resp.Header.Get("Retry-After"); retryAfterStr != "" {
					if parsed, err := strconv.Atoi(retryAfterStr); err == nil {
						retryAfter = parsed
					}
				}
				time.Sleep(time.Duration(retryAfter) * time.Second)
				continue
			}
			// Max retries exhausted, fall through to readJSONResponse which will handle the error
		}

		return readJSONResponse(resp)
	}
	return nil, fmt.Errorf("unreachable")
}

func readJSONResponse(resp *http.Response) (map[string]interface{}, error) {
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response body: %w", err)
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(data))
	}
	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("unmarshal response: %w", err)
	}
	return result, nil
}

// ExtractCards extracts adaptive card content from a message's attachments.
func ExtractCards(message map[string]interface{}) []interface{} {
	attachments, ok := message["attachments"].([]interface{})
	if !ok {
		return nil
	}
	var cards []interface{}
	for _, att := range attachments {
		attMap, ok := att.(map[string]interface{})
		if !ok {
			continue
		}
		if attMap["contentType"] == AdaptiveCardContentType {
			if content, ok := attMap["content"]; ok {
				cards = append(cards, content)
			}
		}
	}
	return cards
}
